ydb/library/yql/minikql/invoke_builtins/mkql_builtins_mul.cpp

```diff
@@ -29,32 +29,79 @@ struct TMul : public TSimpleArithmeticBinary<TLeft, TRight, TOutput, TMul<TLeft,
+            if (right.Get<ui64>() > static_cast<ui64>(std::numeric_limits<i64>::max())) {
+                return NUdf::TUnboxedValuePod();
+            }
+        }
+        i64 lvAbs = (lv > 0) ? lv : -lv;
+        i64 rvAbs = (rv > 0) ? rv : -rv;
+        if (rvAbs != 0 && (std::numeric_limits<i64>::max() / rvAbs < lvAbs)) {
+            return NUdf::TUnboxedValuePod();
+        }
         return IsBadInterval<TOutput>(ret) ? NUdf::TUnboxedValuePod() : NUdf::TUnboxedValuePod(ret);
     }
 
 #ifndef MKQL_DISABLE_CODEGEN
     static Value* Generate(Value* left, Value* right, const TCodegenContext& ctx, BasicBlock*& block)
     {
         auto& context = ctx.Codegen.GetContext();
-        const auto lhs = StaticCast<typename TLeft::TLayout, i64>(
-                GetterFor<typename TLeft::TLayout>(left, context, block), context, block);
-        const auto rhs = StaticCast<typename TRight::TLayout, i64>(
-                GetterFor<typename TRight::TLayout>(right, context, block), context, block);
+        const auto lv = GetterFor<typename TLeft::TLayout>(left, context, block);
+        const auto lhs = StaticCast<typename TLeft::TLayout, i64>(lv, context, block);
+        const auto rv = GetterFor<typename TRight::TLayout>(right, context, block);
+        const auto rhs = StaticCast<typename TRight::TLayout, i64>(rv, context, block);
         const auto mul = BinaryOperator::CreateMul(lhs, rhs, "mul", block);
-        const auto full = SetterFor<typename TOutput::TLayout>(mul, context, block);
-        const auto bad = GenIsBadInterval<TOutput>(mul, context, block);
-        const auto zero = ConstantInt::get(Type::getInt128Ty(context), 0);
-        const auto sel = SelectInst::Create(bad, zero, full, "sel", block);
-        return sel;
+        const auto result = SetterFor<typename TOutput::TLayout>(mul, context, block);
+
+        const auto zero = ConstantInt::get(Type::getInt64Ty(context), 0);
+        const auto lhsAbs = SelectInst::Create(
+                CmpInst::Create(Instruction::ICmp, ICmpInst::ICMP_SGT, lhs, zero, "lhsPos", block),
+                lhs,
+                BinaryOperator::CreateNeg(lhs, "lhsNeg", block),
+                "lhsAbs", block);
+        const auto rhsAbs = SelectInst::Create(
+                CmpInst::Create(Instruction::ICmp, ICmpInst::ICMP_SGT, rhs, zero, "rhsPos", block),
+                rhs,
+                SelectInst::Create(
+                    CmpInst::Create(Instruction::ICmp, ICmpInst::ICMP_EQ, rhs, zero, "rhsZero", block),
+                    ConstantInt::get(Type::getInt64Ty(context), 1),
+                    BinaryOperator::CreateNeg(rhs, "rhsNeg", block),
+                    "rhsNegZero", block),
+                "rhsAbs", block);
+
+        const auto i64Max = ConstantInt::get(Type::getInt64Ty(context), std::numeric_limits<i64>::max());
+        const auto div = BinaryOperator::CreateSDiv(i64Max, rhsAbs, "div", block);
```

Comment:
I can't see an conditional jump over this div (to avoid rvAbs == 0)

Reply:
rhsAbs is set to 1 instead of 0 at line 84
Agreed to add explicit branch to handle zero in rhs to make llvm code similar to scalar code above